contributor-docs/adrs/adr-008-experimental-components.md

```diff
@@ -0,0 +1,77 @@
+- A draft has experimental API, but follows the development conventions of primer/react
+- Components in drafts are accompanied by a design spec and go through the process of accessibility reviews. A component can check multiple boxes in the component lifecycle.
+- The contract with consumers is that you are opting into a component in which we don't have 100% confidence yet.
+
+[^1]: Calling this a prototype and not experimental, because that already means something else in the maturity lifecycle.
+
+&nbsp;
+
+#### Problems with drafts
+
+1. **Incompatible with semver**: Because the expectation with drafts is that we are not 100% confident about the API or structure of a component, we slip in breaking changes to draft components as minor releases to primer/react. On the consumer side, this makes every upgrade, even a patch, a potentially breaking upgrade. So, you are effectively opting out of your semver.
+
+   Upstreaming a component that was built in the context of a product requires re-evaluating the baked-in assumptions and most likely a change in the API to make it useful for other contexts.
+
+1. **Confusion about maturity**: Because of the generic name, "drafts" causes a lot of confusion, especially with "experimental". It is a prototype or experiment, but NOT in the way the [component lifecycle]([https://primer.style/contribute/component-lifecycle]) defines "Experimental" maturity.
+
+1. **Need for experimental candidates**: As more teams build React components and look to upstream reusable components, it is natural to look at drafts as the place to put them. However, the barrier to entry into drafts isn't low (because we still follow design reviews, dev conventions, API reviews, etc.)
+
+   Feature teams using Primer may not be intimately familiar with Primer component lifecycle criteria or have the time and/or knowledge to meet it. Often, feature teams create their own components outside of Primer if they find it too difficult to contribute. This leads to lower Primer adoption and conflicting versions of similar functionality throughout GitHub products.
+
+   We want to enable feature teams to contribute to Primer. We also want to make it clear which contributions haven't been evaluated by our high standards for Primer components.
+
+
+
+&nbsp;
+
+## Decision
+
+1. npm package: "Experimental" components, as defined by the component lifecycle, should not be part of semantically versioned npm package for `@primer/react`.
+
+2. npm package: Each experimental component should be independently published and versioned with semver. Breaking changes should be tagged as major versions. Example: `@github/react-commentbox`
+
+3. The code for experimental components should live in [github/github/modules/react-shared](https://github.com/github/github/tree/master/app/assets/modules/react-shared) and published as individual packages as suggested in point 2.
```

Comment:
I think this will be a huge improvement, thank you for working on this!

I do have a few concerns with placing these components inside gh/gh instead of in their own repo:

1. Tooling
  - Will we be able to setup Storybook and linting with this strategy?
  - Will we be able to setup a staging environment for PRs?
3. Publishing
  - Will it be automatic, and if so how will we set the versions?
  - Are we going to have to use the dotcom deploy queue to publish these packages, and if we don't, how will we ensure that the code gets merged into `master` (due to the deploy-then-merge setup)?
  - How do we ensure that, when code in this directory changes, a new version gets published instead of just merging it to `master`?
5. Usage
  - When code inside the monolith uses these components, will they import directly or will they install the package?
  - If they import directly, how can they take advantage of semver to manage breaking changes?

Personally it feels like a lot of these problems would be solved with a new repository that has its own tooling configured. Then everyone everywhere will depend on the components using the NPM packages. And development would be easier in a smaller repo with a tailored Codespaces setup.

Reply:
All very good questions! I appreciate it!

> 1. Tooling
>   Will we be able to setup Storybook and linting with this strategy?
>   Will we be able to setup a staging environment for PRs?

Yes and yes :) Improving tooling would be the first step if we go forward with this ADR. I'm hoping we can utilise [primer/react-scripts](https://github.com/primer/react-scripts) work for the react-shared

> 2. Publishing
>   Will it be automatic, and if so how will we set the versions?
>   Are we going to have to use the dotcom deploy queue to publish these packages, and if we don't, how will we ensure that the code gets merged into master (due to the deploy-then-merge setup)?
>  How do we ensure that, when code in this directory changes, a new version gets published instead of just merging it to master?

I have some opinions but I don't think I can make that decision, I'm happy to collab on a followup ADR in github/github for these questions with someone who works on github/github regularly.

> 3. Usage
> When code inside the monolith uses these components, will they import directly or will they install the package?

I want to say yes, that's how it is setup currently and has the lowest barrier to contribution.

> If they import directly, how can they take advantage of semver to manage breaking changes?

There is a risk of introducing breaking changes for non-monolith projects and not realising it. I haven't spent enough time thinking about that, we should address it in the ADR about publishing react-shared

Reply:
Thanks!

It feels like it's going to be very hard to avoid a complex situation where we have some code importing and some using the published package. It will be very challenging to ensure that importing code (teams inside the monolith) must always publish all the changes they make, particularly if there is no project actively using the published package version. It will also be very difficult for non-monolith teams to make changes if there are in-monolith teams importing the code, because they will have to address all breaking changes in other projects before they can publish anything.

Additionally - if the components are built inside the monolith, what's to stop them from importing code that also exists in the monolith, outside the react-shared folder? If they do that, how will we publish the code without blowing up the package to include _all_ monolith JS/TS code?

Maybe all these problems have solutions through linting and tooling and policies, but it seems like a lot of complexity - maybe the 'low barrier to entry' of an in-monolith package is really an illusion.

Reply:
> It feels like it's going to be very hard to avoid a complex situation where we have some code importing and some using the published package. It will be very challenging to ensure that importing code (teams inside the monolith) must always publish all the changes they make, particularly if there is no project actively using the published package version.

(This needs more thought, but...) Would it be fair to say if it's merged into main/master, we should publish it for other teams to use? We could make changelog required for react-shared ([just like primer](https://github.com/primer/react/blob/main/.github/workflows/check_for_changeset.yml)) and publish on merge.

cc @joshblack has some experience with this.

> It will also be very difficult for non-monolith teams to make changes if there are in-monolith teams importing the code, because they will have to address all breaking changes in other projects before they can publish anything.

This is an issue with experimental components in general, no matter where you keep them 😓 Because, there is no commitment to stability, one team can make changes without the knowledge (or testing) that it breaks in a different non-monolith product.

> Additionally - if the components are built inside the monolith, what's to stop them from importing code that also exists in the monolith, outside the react-shared folder? If they do that, how will we publish the code without blowing up the package to include all monolith JS/TS code?
> Maybe all these problems have solutions through linting and tooling and policies

This is definitely a linting opportunity.


> Maybe all these problems have solutions through linting and tooling and policies, but it seems like a lot of complexity - maybe the 'low barrier to entry' of an in-monolith package is really an illusion.

That's a good point! The barrier to entry is relative to primer/react, even then, the barrier is lowered for devs working on the monolith, but not as much for non-monolith projects like memex. (this is something i call out in the ADR, we'll have to find out how bad it is)

Another option (mentioned in [options considered in this ADR](https://github.com/primer/react/blob/adr-experimental-package/contributor-docs/adrs/adr-008-experimental-components.md#other-options-considered)) was to keep experimental components in a repository + package outside the monolith (example: github.com/github/react-shared) which adds some friction for monolith but solves the process/governance problem by creating the same environment for monolith and non-monolith projects. Ultimately, we decided the best tradeoff is to optimise for the monolith were react-shared lives today and then add tooling + processes to reduce friction for non-monolith projects.

Reply:
Something else to consider as well is if we choose the non-monolith route, we would import this package into gh/gh and A) guarantee that all parts of GitHub are working with the same code and B) will de-dupe any package imports of this package.

Reply:
>Something else to consider as well is if we choose the non-monolith route, we would import this package into gh/gh and A) guarantee that all parts of GitHub are working with the same code and B) will de-dupe any package imports of this package.

That's exactly my thinking as well - it has the huge benefit of enforcing that everyone uses the code the same way, which could save us some serious headaches.

Reply:
> Are we going to have to use the dotcom deploy queue to publish these packages, and if we don't, how will we ensure that the code gets merged into master (due to the deploy-then-merge setup)?

IMO it would be best to explore what the build, deploy and merge story would be for a new package in dotcom before committing to that vs. an external package. To me, it is less than ideal currently for Primer Rails Experimental components, as we adhere to all the same mechanics as everyone else working in dotcom. The main issues with that for component development are:
- No preview (lab) deploys until tests are passing and PR has been approved. Maybe this would be different for a Storybook deploy that doesn't use production data?
- Builds take forever and often fail for things unrelated to a given PR
- Need to be present for merging (trains) even if the thing you're shipping isn't used anywhere. 

I suppose you would need to go through all those steps to bump a package version as well. So maybe it doesn't matter?

Reply:
> (This needs more thought, but...) Would it be fair to say if it's merged into main/master, we should publish it for other teams to use? We could make changelog required for react-shared ([just like primer](https://github.com/primer/react/blob/main/.github/workflows/check_for_changeset.yml)) and publish on merge.
>
> cc @joshblack has some experience with this.


Hey! Definitely think this would be super helpful, especially if it's intended for other teams outside the monolith to use this work.

If it helps, here's a rough overview of a setup I've used in the past:

| Package | Owner | Version Policy | Release cadence | Repo |
| :------ | :---- | :------------- | :-------------- | :------- |
| `@ds/styles` | Core | semver | release train, biweekly | DS Monorepo |
| `@ds/react` | Core | semver | release train, biweekly | DS Monorepo |
| `@ds-incubator/react` | Maintainer(s) of contributed code | pre-1.0 | publish-on-merge | Incubator monorepo |
| `@ds-incubator/<component>` | Maintainer(s) of contributed code | pre-1.0 | publish-on-merge | Incubator monorepo |
| `@local-system/react` | Local system core | semver | publish-on-merge | Local system monorepo |
| `@local-system/<component>` | Maintainer(s) | semver or pre-1.0 | release train, as-needed, or publish-on-merge | Either in monorepo or standalone repo |

Generally:

- Each publish would generate a changelog
- The components of each package had some form of documentation site (designsystem.com/local-system) or was flagged as experimental in a peer docs site, like storybook
- A good number of teams chose a publish-on-merge strategy for faster iterations
- Pre-1.0 was for experimentation and alignment, the road to 1.0 generally involved a commitment to maintenance by core or a community member along with stabilization of the API
- "Monorepos" generally referred to projects that:
  - Published multiple packages (typically via lerna)
  - Tended to have one docs/storybook site
  - Used Yarn Workspaces

Incubator was used for opportunities where several teams may want to share some work and could use shared tooling (linting, formatting, storybook) and processes (releases, deploys) in order to focus on building components.

Reply:
While we don't need to make the decisions about publishing react-shared in this ADR, it's been really useful to collect concerns and challenges before approving this ADR to make sure we're not running into a wall.

Next steps: We need an ADR in gh/gh for publishing react-shared because there are other folks that should be involved. There is a big chunk of work between "Accepting" the current ADR (008) and "Adopting" it.

If we decide to move react-shared out of gh/gh, we can revisit this ADR and update the decision, all part of the process.

Reply:
Are there any current examples of packages actively getting published from inside the gh/gh monolith repository? If so, that might alleviate some of the concern. If not, that might be a warning sign.